### PR TITLE
feat(payment-service): add subscription payments

### LINKED ISSUE #359: Recurring/Subscription Payment Service
Ability to integrate recurring payments

### PR DESCRIPTION
**Pull request by sumiter92**
## Description

Adds subscription feature in payment service that supports configurable multiple payment gateways.

Fixes #359 

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [ ] Performed a self-review of my own code
- [ ] npm test passes on your machine
- [ ] Code conforms with the style guide

### CODE REVIEW

File: services/payment-service/src/controllers/transactions.controller.ts [hunk 1]
```diff
@@ -330,4 +333,147 @@ export class TransactionsController {
...
+    const hostUrl = this.req.get('host');
+    const hostProtocol = this.req.protocol;
+    this.req.query.method = newSubscription.paymentMethod;
+    return this.res.redirect(
+      redirectStatusCode,
+      `${hostProtocol}://${hostUrl}/transactions/subscriptionid/${newSubscription.id}?method=${newSubscription.paymentMethod}`,
+    );
+  }
+
+  @get(`/transactions/subscriptionid/{id}`)
```
Comment (samarpan-b) [inline services/payment-service/src/controllers/transactions.controller.ts:384]: incorrect REST format for url.

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: services/payment-service/src/controllers/transactions-subscriptions.controller.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+          type: 'object',
+        },
+      },
+    },
+  })
+  async subscriptionTransactionsPay(
+    @param.path.string('id') id: string,
+  ): Promise<unknown> {
+    const Subscription = await this.subscriptionRepository.findById(id);
+    const templates = await this.templatesRepository.find({
```
Comment (akshatdubeysf) [inline services/payment-service/src/controllers/transactions-subscriptions.controller.ts:44]: Why not use findOne?

File: services/payment-service/src/controllers/transactions-subscriptions.controller.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      where: {
+        paymentGatewayId: Subscription?.paymentGatewayId,
+        name: TemplateName.Create,
+      },
+    });
+    const paymentTemplate = templates[0]?.template;
+    return this.res.send(
+      await this.gatewayHelper.subscriptionCreate(
```
Comment (akshatdubeysf) [inline services/payment-service/src/controllers/transactions-subscriptions.controller.ts:52]: incase paymentTemplate is undefined?

File: services/payment-service/src/controllers/subscriptions.controller.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+          }),
+        },
+      },
+    })
+    subscriptions: Subscriptions,
+  ): Promise<Subscriptions> {
+    return this.subscriptionsRepository.create(subscriptions);
+  }
+
+  @get('/subscriptions/count')
```
Comment (akshatdubeysf) [inline services/payment-service/src/controllers/subscriptions.controller.ts:52]: authorization and authentication?

Comment (sumiter92) [reply services/payment-service/src/controllers/subscriptions.controller.ts:59]: added

**Pull request merged**